### PR TITLE
`opt_frozen_dataclass`: Enable hashing with -O

### PR DESCRIPTION
**Pull request by inducer**
Inspired by https://github.com/inducer/pytato/pull/563.

cc @matthiasdiener (what do you think?)

### CODE REVIEW

**Comment by matthiasdiener** [comment]
Dunno, seems risky 😬

**Comment by matthiasdiener** [comment]
Maybe replacing `unsafe_hash: bool = False,` with `unsafe_hash: bool = None,` could be an option?

**Comment by inducer** [comment]
> Dunno, seems risky 😬

I'd argue it's needed for correctness. Otherwise, these classes are hashable when running without `-O` and not hashable when running with `-O`, which is inconsistent. They're supposed to be semantically frozen classes, as the name suggests. I'll add a comment as to why that's there.

**Comment by inducer** [comment]
> Maybe replacing `unsafe_hash: bool = False,` with `unsafe_hash: bool = None,` could be an option?

Sure, I could see that.

**Comment by inducer** [comment]
Paging @alexfikl for a third set of eyes: Is all this just a terrible idea and not worth it?

File: pytools/__init__.py [hunk 1]
```diff
@@ -3000,16 +3000,38 @@ def opt_frozen_dataclass(
...
+        and vice versa. To ensure frozen-ness is applied predictably in all
+        scenarios (mainly :data:`__debug__` on and off), it is strongly recommended
+        that all dataclasses inheriting from ones with this decorator *also*
+        use this decorator. There are no run-time checks to make sure of this.
+
     .. versionadded:: 2024.1.18
     """
     def map_cls(cls: type[T]) -> type[T]:
+        # This ensures that the resulting dataclass is hashable with and without
+        # __debug__, unless the user overrides unsafe_hash.
```
Comment (matthiasdiener) [inline pytools/__init__.py:3019]:
```suggestion
        # __debug__, unless the user overrides unsafe_hash, or provides their own __hash__ method.
```

File: pytools/__init__.py [hunk 2]
```diff
@@ -3000,16 +3000,38 @@ def opt_frozen_dataclass(
...
+        if (eq
+                and not __debug__
+                and unsafe_hash is None
+                and "__hash__" not in cls.__dict__):
+            loc_unsafe_hash = True
+        else:
+            loc_unsafe_hash = False
+
```
Comment (matthiasdiener) [inline pytools/__init__.py:3027]: I think this still doesn't allow the user to pass `unsafe_hash=True`?

File: pytools/__init__.py [hunk 3]
```diff
@@ -3000,16 +3000,38 @@ def opt_frozen_dataclass(
...
     def map_cls(cls: type[T]) -> type[T]:
+        # This ensures that the resulting dataclass is hashable with and without
+        # __debug__, unless the user overrides unsafe_hash.
+        if (eq
+                and not __debug__
+                and unsafe_hash is None
+                and "__hash__" not in cls.__dict__):
+            loc_unsafe_hash = True
+        else:
+            loc_unsafe_hash = False
```
Comment (matthiasdiener) [inline pytools/__init__.py:3026]: maybe:

```suggestion
            loc_unsafe_hash = unsafe_hash
```

?

Comment (inducer) [reply pytools/__init__.py]: Addressed, thanks!

File: pytools/__init__.py (hunk 2 above)
Comment (inducer) [reply pytools/__init__.py]: Addressed, thanks!

File: pytools/__init__.py (hunk 1 above)
Comment (inducer) [reply pytools/__init__.py]: Addressed, thanks!

**Comment by matthiasdiener** [comment]
Side note: [this table](https://github.com/python/cpython/blob/3.13/Lib/dataclasses.py#L121) is really helpful when thinking through this.

File: pytools/__init__.py [hunk 4]
```diff
@@ -3000,16 +3000,41 @@ def opt_frozen_dataclass(
     this decorator avoid when the interpreter runs with "optimization"
```
Comment (matthiasdiener) [inline pytools/__init__.py:3000]:
```suggestion
    this decorator avoids when the interpreter runs with "optimization"
```

File: pytools/__init__.py [hunk 5]
```diff
@@ -3000,16 +3000,41 @@ def opt_frozen_dataclass(
...
     enabled.
 
+    The resulting dataclass supports hashing unless *eq* is set to *False*,
+    if *unsafe_hash* is left at the default or set to *True*.
```
Comment (matthiasdiener) [inline pytools/__init__.py:3004]:
```suggestion
    The resulting dataclass supports hashing, even when it is not actually frozen,
    unless *eq* is set to *False*, if *unsafe_hash* is left at the default or set to *True*.
```

**Review by matthiasdiener — APPROVED** [review]
Besides the minor wording comments, this LGTM. Thanks!

**Comment by alexfikl** [comment]
> Paging @alexfikl for a third set of eyes: Is all this just a terrible idea and not worth it?

Hm, I haven't followed this much, but if I understand correctly.. we want to turn off `frozen` when in `-O` mode because it's slower, but that also turns off hashing by default, right? And this is meant to sneakily add back hashing as if the class was frozen?

**Comment by inducer** [comment]
Thanks all for your help in getting this into shape. I'll do another release as soon as this is in.

**Comment by inducer** [comment]
https://pypi.org/project/pytools/2024.1.19/

**Pull request merged**